libs/gi/page-archive/src/TabWeapon.tsx

```diff
@@ -0,0 +1,203 @@
+import StarRoundedIcon from '@mui/icons-material/StarRounded'
+import type { Palette } from '@mui/material'
+import {
+  Box,
+  CardContent,
+  InputAdornment,
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableRow,
+  TextField,
+  ToggleButton,
+  ToggleButtonGroup,
+  Typography,
+} from '@mui/material'
+import { memo, useDeferredValue, useMemo, useState } from 'react'
+import { WeaponView } from './WeaponView'
+const rarities = [5, 4, 3, 2, 1] as const
+export default function TabWeapon() {
+  const [rarityFilter, setRarityFilter] = useState([...rarities])
+  const [weaponTypeFilter, setWeaponTypeFilter] = useState([
+    ...allWeaponTypeKeys,
+  ])
+  const handleRarity = handleMultiSelect([...rarities])
+  const handleType = handleMultiSelect([...allWeaponTypeKeys])
+  const [searchTerm, setSearchTerm] = useState('')
+  const searchTermDeferred = useDeferredValue(searchTerm)
+  const weaponKeys = useMemo(() => {
+    return allWeaponKeys.filter(
+      (wKey) => {
+        const { rarity, weaponType } = getWeaponStat(wKey)
+        if (!rarityFilter.includes(rarity)) return false
+        if (!weaponTypeFilter.includes(weaponType)) return false
+        const setKeyStr = i18n.t(`weaponNames_gen:${wKey}`)
+        if (
+          searchTermDeferred &&
+          !setKeyStr
+            .toLocaleLowerCase()
+            .includes(searchTermDeferred.toLocaleLowerCase())
+        )
+          return false
+        return true
+      },
+      [rarityFilter]
+    )
+  }, [rarityFilter, searchTermDeferred, weaponTypeFilter])
+  return (
+    <Box>
+      <CardContent sx={{ display: 'flex', gap: 2 }}>
+        <ToggleButtonGroup value={rarityFilter}>
+          {rarities.map((r) => (
+            <ToggleButton
+              key={r}
+              value={r}
+              onClick={() => setRarityFilter((old) => handleRarity(old, r))}
+            >
+              <ColorText color={`rarity${r}` as keyof Palette}>
+                <StarRoundedIcon sx={{ verticalAlign: 'text-top' }} />
+              </ColorText>
+            </ToggleButton>
+          ))}
+        </ToggleButtonGroup>
+        <ToggleButtonGroup value={weaponTypeFilter}>
+          {allWeaponTypeKeys.map((wt) => (
+            <ToggleButton
+              key={wt}
+              value={wt}
+              onClick={() => setWeaponTypeFilter((old) => handleType(old, wt))}
+            >
+              <ImgIcon src={imgAssets.weaponTypes?.[wt]} size={2} />
+            </ToggleButton>
+          ))}
+        </ToggleButtonGroup>
+        <TextField
+          fullWidth
+          variant="outlined"
+          value={searchTerm}
+          onChange={(e) => setSearchTerm(e.target.value)}
+          InputProps={{
+            startAdornment: (
+              <InputAdornment position="start">
+                <SearchIcon />
+              </InputAdornment>
+            ),
+          }}
+        />
+      </CardContent>
+      <Table>
+        <TableHead>
```

Comment:
Is there any way to do the infinite scrolling pagination? Took me literally 5s to load the weapon archive, i assume it was downloading all the images and stuff. Alternatively, if we could load the header of the table with the filtering, then display a spinner on the table body while it loads, it would look a bit cleaner